lib/WUI/http/upload_state.cpp

```diff
@@ -0,0 +1,649 @@
+    }
+    /*
+     * Access the internal buffer.
+     *
+     * This can be used to recycle the buffer for different purposes (to save
+     * memory) while not being in used. The buffer or the functionality of
+     * looking for something are mutually exclusive (use one or the other, not
+     * both at the same time).
+     */
+    array<char, MAX_SUBTOKEN> &borrow_buf() {
+        return buffer;
+    }
+};
+
+class UploadState {
+private:
+    /*
+     * This helps tracking if we start a new token or if this is a continuation
+     * of the same one.
+     */
+    // Note: Use of 'enum class' generates a warning at the place of the
+    // bitfield usage below, even though it is big enough.
+    enum TokenType {
+        NoToken,
+        HeaderField,
+        HeaderValue,
+        DataToken,
+        // Warning: Check the bitfield sizes if adding values.
+    };
+
+    /*
+     * The states of the automaton.
+     */
+    enum State {
+        NoState,
+        CheckHeaderIsCDisp,
+        IgnoredHeader,
+        CDispHeader,
+        ReadString,
+        ReadPartName,
+        Data,
+        Done,
+        // Warning: Check the bitfield sizes if adding values.
+    };
+
+    /*
+     * Whenever we found a part of a known name, we record it by this.
+     *
+     * Anything unknown or before finding the name, it is Unknown and just
+     * ignored.
+     */
+    enum Part {
+        Unknown,
+        Print,
+        File,
+        // Possibly others in the future
+        // Warning: Check the bitfield sizes if adding values.
+    };
+
+    /**
+     * The handlers.
+     *
+     * Not owned by us.
+     */
+    HttpHandlers *handlers;
+
+    unique_ptr<multipart_parser, void (*)(multipart_parser *)> multiparter;
+
+    /*
+     * When reading a string separated by quotes, this points to the current
+     * poisition and an end of a buffer to store it. This makes it possible to
+     * point the reading of characters to different buffers depending on what
+     * the string means.
+     *
+     * Relevant for the ReadString and ReadPartName states.
+     *
+     * When both are pointing to the same location, no more characters are read
+     * and they are simply thrown away. The special-case of both pointing to
+     * NULL is a /dev/null buffer for skipping until the end of a string that's
+     * of no interest to us.
+     */
+    char *string_dst_pos = nullptr;
+    const char *string_dst_end = nullptr;
+
+    // A buffer for accumulating strings. See above.
+    Accumulator accumulator;
+
+    TokenType type : 2;
+    State state : 3;
+    Part part : 2;
+    bool have_valid_filename : 1;
+    /*
+     * The gcode_start hook has been called and we need to eventually call the finish hook too.
+     *
+     * This may get reset back to false in case the finish hook is called or
+     * when there's an error from hook (the hooks should know they erorred out
+     * and clean up their state as needed).
+     */
+    bool init_done : 1;
```

Comment:
Have you checked that the compiler is capable of reordering the member variables/bits to minimze the wasted mem space? And, yes, I read your intro design description, that you "hope" for it.

Reply:
I haven't and I don't think there's a reasonably easy way to do so.

What I'm sure of is, if I put it into a header file, the compiler is denied even the chance, because order of variables in structs is _guaranteed_ in C/C++ and if it's in a header file, there might be some other code that relies on it. By hiding the implementation, I'm giving the compiler more freedom, because it now knows „nobody is watching“, so it can „break the rules“. I would _like_ it to do the optimization with optimizing the members (for example, I'm pretty sure not all combination of states/tokens/parts is possible), but there's likely a huge number of other optimizations it may unlock (or not).

I'm still trying to provide the bit fields and order the members according to their size/alignment so even the „right“ representation would be somewhat OK.

Reply:
OK, good enough for me :wink: